Assets/Scripts/Behaviours/Vehicles/Vehicle_Spawning.cs

```diff
@@ -618,5 +635,150 @@ void SetupDoorsHingeJoints()
+                if (numComps > 0)
+                    return anVariations[UnityEngine.Random.Range(0, numComps)];
+            }
+
+            return -1;
+        }
+
+        private bool IsValidCompRule(int nRule)
+        {
+            //    TODO add weather checking, when weather is implemented.
+            Boolean isRainingNow = false;
+
+            return (nRule != (int)VehicleComponentsRules.ONLY_WHEN_RAINING)
+                || isRainingNow
+            ;
+        }
+
+        private int ChooseComponentInternal(int rule, int comps)
+        {
+            int component = -1;
+            if (rule == (int)VehicleComponentsRules.ALLOW_ALWAYS || 
+                rule == (int)VehicleComponentsRules.ONLY_WHEN_RAINING)
+            {
+                int iNumComps = CountCompsInRule(comps);
+                int rand = UnityEngine.Random.Range(0, iNumComps);
+                component = (comps >> (4 * rand)) & 0xF;
+            }
+            else if (rule == (int)VehicleComponentsRules.MAYBE_HIDE)
+            {
+                int iNumComps = CountCompsInRule(comps);
+                int rand = UnityEngine.Random.Range(-1, iNumComps);
+                if (rand != -1)
+                    component = (comps >> (4 * rand)) & 0xF;
+            }
+            else if (rule == (int)VehicleComponentsRules.FULL_RANDOM)
+            {
+                component = UnityEngine.Random.Range(0, 5);
+            }
+
+            return component;
+        }
+
+        int CountCompsInRule(int comps)
+        {
+            int result = 0;
+            while (comps != 0)
+            {
+                if ((comps & 0xF) != 0xF)
+                    ++result;
+
+                int a = comps / 16;
+                int b = comps >> 4;
+
+                comps >>= 4;
+            }
+
+            return result;
+        }
+
+        int GetListOfComponentsNotUsedByRules(int numExtras, int[] outList)
+        {
+            int[] iCompsList = new int[]{ 0, 1, 2, 3, 4, 5 };
+
+            VehicleDef.CompRulesUnion comps = Definition.CompRules;
+
+            if (comps.nExtraARule != 0 && IsValidCompRule(comps.nExtraARule))
+            {
+                if (comps.nExtraARule == (int)VehicleComponentsRules.FULL_RANDOM)
+                    return 0;
+
+                if (comps.nExtraA_comp1 != 0xF)
+                    iCompsList[comps.nExtraA_comp1] = 0xF;
+
+                if (comps.nExtraA_comp2 != 0xF)
+                    iCompsList[comps.nExtraA_comp2] = 0xF;
+
+                if (comps.nExtraA_comp3 != 0xF)
+                    iCompsList[comps.nExtraA_comp3] = 0xF;
+            }
+
+            if (comps.nExtraBRule != 0 && IsValidCompRule(comps.nExtraBRule))
+            {
+                if (comps.nExtraBRule == (int)VehicleComponentsRules.FULL_RANDOM)
+                    return 0;
+
+                if (comps.nExtraB_comp1 != 0xF)
+                    iCompsList[comps.nExtraA_comp1] = 0xF;
```

Comment:
should these array indexes use `B` ? eg: `nExtraB_comp1`, `nExtraB_comp2`, `nExtraB_comp3`

Reply:
Yes, using A makes no sense here, copy-paste issue.